src/metabase/models/notification.clj

```diff
@@ -391,3 +428,29 @@
               handler-id (t2/insert-returning-pk! :model/NotificationHandler handler)]
           (t2/insert! :model/NotificationRecipient (map #(assoc % :notification_handler_id handler-id) recipients))))
       instance)))
+
+(def ^:private notification-update-spec
+  {:model        :model/Notification
+   ;; a function that takes a row and returns a map of the columns to compare
+   :compare-row  #(select-keys % [:active])
+   :nested-specs {:payload       {:model        :model/NotificationCard
+                                  :compare-row  #(select-keys % [:send_condition :send_once])}
+                  :subscriptions {:model        :model/NotificationSubscription
+                                  ;; the foreign key column in the nested model with respect to the parent model
+                                  :fk-column    :notification_id
+                                  :compare-row  #(select-keys % [:notification_id :type :event_name :cron_schedule])
+                                  ;; whether this nested model is a sequentials with respect to the parent model
+                                  :multi-row?   true}
+                  :handlers      {:model        :model/NotificationHandler
+                                  :fk-column    :notification_id
+                                  :compare-row  #(select-keys % [:notification_id :channel_type :channel_id :template_id :active])
+                                  :multi-row?   true
+                                  :nested-specs {:recipients {:model       :model/NotificationRecipient
+                                                              :fk-column   :notification_handler_id
+                                                              :compare-row #(select-keys % [:notification_handler_id :type :user_id :permissions_group_id :details])
+                                                              :multi-row?  true}}}}})
+
+(defn update-notification!
```

Comment:
the 2nd model is unnatural because system event doesn't have an equivalent table

Reply:
Can you refresh my memory on why NotificationCard needs to be inserted first? It sounds like a M2M join table, so why is the relationship inverted?

It feels a bit strange for the spec nesting to be different to the database structure.

Maybe you could just define both as top level specifications, with a keyword instead of a map for references to non "private" specs?

```
- notification_card
  - *notification*
- notification
  - handlers
  - subscriptions
```

Reply:
@qnkhuat so the fact that `NotificationSubscription` also needs to be created before `Notification` is fine, but for `NotificationCard` it's a problem?

Reply:
maybe I missed word somewhere but `NotificationSubscription` doesn't need to be created before, in fact it has to be created after.

Why does `notification_card` need to be created first? It is because of `notification.payload_id` implicitly is a FK to `notification_card.id`

Reply:
Ah yeah I agree this is not a good enough reason to flip the more natural hierarchy.

Post-processing might be the cleanest solution here, as ugly as that is.

Reply:
What about using an entity ID for the relationship instead?

Reply:
Could you guys clarify what do you mean by post-processing? if any I think we need preprocessing right? in this case we'need to pre-insert notification-card, so that we can use spec to update properly

Reply:
I mean inserting the notification first, without the payload id, then updating the id afterwards. A bit awkward with the defensive hooks, but I think this is more intuitive and generalizable than reshuffling the creation order.

Reply:
ah okay. but that's problem for another PR where we replace `create-nofication!` using this spec. are there anything we need to resolve before moving forward here?

Reply:
nothing blocking. you started this thread, feel free to resolve it 😄